server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -553,22 +564,13 @@ private static Map<String, Mapper> buildMergedMappers(
                     // If we're merging template mappings when creating an index, then a field definition always
                     // replaces an existing one.
                     if (reason == MergeReason.INDEX_TEMPLATE) {
-                        merged = mergeWithMapper;
+                        mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper);
                     } else {
-                        merged = mergeIntoMapper.merge(mergeWithMapper, objectMergeContext);
+                        mergedMappers.put(mergeWithMapper.simpleName(), mergeIntoMapper.merge(mergeWithMapper, objectMergeContext));
```

Comment:
Is it necessary to move where the put happens (within each if branch compared to at the end)? I am trying to figure out if that is just noise in the diff. If possible, I'd like to keep things as they were as much as possible.

Reply:
I've tried again, but the recursive nature of this, combined with the fact that we want to limit fields here makes this challenging.

If we only add the merged mappers to the map at the end of the method, and if we decrement the number of fields at the end, we would first consume the budget for the child mappers so that when the recursion ends and we're putting the mappers into the map, the child mappers may have consumed all of the budget so that we can't add the parent objects in.

The intention of the `MapperMergeContext.addFieldIfPossible` was to keep the act of checking the limit, adding the mapper to the map, and decrementing the budget as close to one another as possible. Otherwise, it can be easy to miss a step.